internal/rpc/clients.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpc
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"open-match.dev/open-match/internal/config"
+)
+
+var (
+	clientLogger = logrus.WithFields(logrus.Fields{
+		"app":       "openmatch",
+		"component": "client",
+	})
+)
+
+// ClientParams contains the connection parameters to connect to an Open Match service.
+type ClientParams struct {
+	Hostname       string
+	Port           int
+	TrustedKeyPath string
+}
+
+func (p *ClientParams) usingTLS() bool {
+	return len(p.TrustedKeyPath) > 0
+}
+
+// GRPCClientFromConfig creates a gRPC client connection from a configuration.
+func GRPCClientFromConfig(cfg config.View, prefix string) (*grpc.ClientConn, error) {
+	clientParams := &ClientParams{
+		Hostname: cfg.GetString(prefix + ".hostname"),
+		Port:     cfg.GetInt(prefix + ".grpcport"),
+	}
+
+	// If TLS support is enabled in the config, fill in the trusted key for decrpting server certificate.
+	if cfg.GetBool("tls.enabled") {
+		clientParams.TrustedKeyPath = cfg.GetString("tls.trustedKeyPath")
+	}
+
+	return GRPCClientFromParams(clientParams)
+}
+
+// GRPCClientFromParams creates a gRPC client connection from the parameters.
+func GRPCClientFromParams(params *ClientParams) (*grpc.ClientConn, error) {
+	address := fmt.Sprintf("%s:%d", params.Hostname, params.Port)
+
+	grpcOptions := []grpc.DialOption{}
+
+	if params.usingTLS() {
+		creds, err := clientCredentialsFromFile(params.TrustedKeyPath)
+		if err != nil {
+			clientLogger.WithError(err).Error("failed to get transport credentials from file.")
+			return nil, errors.WithStack(err)
+		}
+		grpcOptions = append(grpcOptions, grpc.WithTransportCredentials(creds))
+	} else {
+		grpcOptions = append(grpcOptions, grpc.WithInsecure())
+	}
+
+	return grpc.Dial(address, grpcOptions...)
+}
+
+// HTTPClientFromConfig creates a HTTP client from from a configuration.
+func HTTPClientFromConfig(cfg config.View, prefix string) (*http.Client, string, error) {
+	clientParams := &ClientParams{
+		Hostname: cfg.GetString(prefix + ".hostname"),
+		Port:     cfg.GetInt(prefix + ".httpport"),
+	}
+
+	// If TLS support is enabled in the config, fill in the trustedKeyPath
+	if cfg.GetBool("tls.enabled") {
+		clientParams.TrustedKeyPath = cfg.GetString("tls.trustedKeyPath")
```

Comment:
You'll want to validate that the file exists and can be read here and return a reasonable error otherwise.